shotover/src/hot_reload_client.rs

```diff
@@ -0,0 +1,207 @@
+//This module gives client-side implementation for socket handoff as part of hot reloading
+//Client will connect to existing shotovers and requests for FDs
+use crate::hot_reload::{Request, Response};
+use crate::json_parsing::read_json;
+use anyhow::{Context, Result};
+use std::time::Duration;
+use tokio::io::{AsyncWriteExt, BufReader};
+use tokio::net::UnixStream;
+use tokio::time::timeout;
+use tracing::{debug, info};
+
+pub struct UnixSocketClient {
+    socket_path: String,
+    timeout_duration: Duration,
+}
+
+impl UnixSocketClient {
+    pub fn new(socket_path: String) -> Self {
+        Self {
+            socket_path,
+            timeout_duration: Duration::from_secs(30),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout_duration = timeout;
+        self
+    }
+
+    pub async fn send_request(&self, request: Request) -> Result<Response> {
+        info!("Connecting to hot reload server at: {}", self.socket_path);
+
+        let result = timeout(self.timeout_duration, self.send_request_inner(request)).await;
+
+        match result {
+            Ok(response) => response,
+            Err(_) => Err(anyhow::anyhow!(
+                "Hot reload request timed out after {:?}",
+                self.timeout_duration
+            )),
+        }
+    }
+
+    async fn send_request_inner(&self, request: Request) -> Result<Response> {
+        // Connect to server
+        let stream = UnixStream::connect(&self.socket_path)
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to connect to hot reload server at: {}",
+                    self.socket_path
+                )
+            })?;
+
+        let (reader, mut writer) = stream.into_split();
+        let mut reader = BufReader::new(reader);
+
+        // Send request
+        let request_json =
+            serde_json::to_string(&request).context("Failed to serialize request")?;
+
+        debug!("Sending request: {}", request_json);
+        writer
+            .write_all(request_json.as_bytes())
+            .await
+            .context("Failed to send request")?;
+
+        // Read response
+        let response: Response = read_json(&mut reader).await?;
+        debug!("Received response: {:?}", response);
+
+        Ok(response)
+    }
+}
+
+/// Request listening sockets from an existing Shotover instance during hot reload
+pub async fn request_listening_sockets(socket_path: String) -> Result<()> {
```

Comment:
I'm thinking lets give this a higher level name like `perform_hot_reloading` that way the runner.rs logic is less concerned with what goes on during the hot reload process.